functional-samples/sample.dnr-rule-manager/manager.js

```diff
@@ -0,0 +1,199 @@
+const ruleItemTemplate = document.getElementById('ruleItemTemplate');
+const rulesList = document.getElementById('rulesList');
+const addRuleButton = document.getElementById('addRuleButton');
+const viewRuleButton = document.getElementById('viewRuleButton');
+
+async function getNextRuleId() {
+  const rules = await chrome.declarativeNetRequest.getDynamicRules();
+  return Math.max(...rules.map((rule) => rule.id)) + 1;
+}
+
+async function refresh() {
+  const rules = await chrome.declarativeNetRequest.getDynamicRules();
+  renderRules(rules);
+}
+
+refresh();
+
+addRuleButton.addEventListener('click', () => {
+  appendRuleItem(rulesList, {
+    id: 'NEW',
+    conditionType: 'urlFilter',
+    conditionValue: '',
+    caseSensitive: true
+  });
+});
+
+viewRuleButton.addEventListener('click', async () => {
+  const rules = await chrome.declarativeNetRequest.getDynamicRules();
+  const rulesString = JSON.stringify(rules, null, 2);
+  const newWindow = window.open();
+  newWindow.document.write(
```

Comment:
Can we find a way to do this by constructing an element vs. writing HTML (similar to wanting to avoid innerHTML)?